### ISSUE TITLE
feature: iterable Scanner

### ISSUE THREAD

**Issue by monperrus** [INRIA/spoon#1077]
spin-off of #1070 

**Idea**: an iterable scanner.

we could then have 

    interface CtElement extends Iterator<CtElement>

so as to easily go over the children elements of an element:

    for (CtElement el : klass) { // do something }

The `CtIterableScanner` class will also be useful for some advanced queries.

It can probably be generated out of CtScanner directly, as we do for Clone, Replace and Equal implementation classes.

**Comment by retuxx** [comment]
You need to define the iteration order, is it pre-, in-, or postfix?

**Comment by pvojtechovsky** [comment]
It is still true that I do not need iterable Scanner for queries, so do not understand this comment as: "Yes it makes sense to do it, because Pavel needs it too". 
But if you think that iterable scanner would be useful, then here is one more idea:
Do not implement just `Iterator` (which has method `next`, but implement these model iterating methods:
* nextChild
* nextSibling
* nextParent
* ?
So the client can drive the way of iterating through the spoon model.

**Comment by pvojtechovsky** [comment]
One note: There are two strategies how to access the data: [push and pull](http://stackoverflow.com/questions/5792966/data-pull-vs-push-oop-approach). 

> Pushing : The advantage of pushing is that **you know your data** and you know what you are pushing. No component knows (or should not know) the data better than the component who owns the data, which will theoretically **imply a better design and a more robust system**.

> Pulling : The only advantage I see in pulling approach is the component who pulls exactly knows when it should pull. It can initiate the conversation when it needs the data and no component knows (or should not know) when the data is needed than the component who needs it.

It depends on situation which one is better - as usually. 

Current CtScanner is implemented using **push** strategy.
The Iterable Scanner will use **pull** strategy.

I do not know your plans with Iterable scanner, but I vote for keeping and future supporting of origin CtScanner, because in some (in case of spoon ... most?) scenarios CtScanner will be much better then iterable scanner.

Second note: the pull and push are incompatible like fire and water. If spoon will support both strategies, then it will probably end with duplicate implementation of all data accessing algorithms. So it increases maintenance cost of the spoon library. May be I am wrong, but I wanted to warn you.

Question: Could you please describe your plans with iterable scanner? Which use cases of spoon clients are covered by this new feature?

**Comment by monperrus** [comment]
> I vote for keeping and future supporting of origin CtScanner

We will keep it. CtIterableScanner will be a separate class. Note that in Spoon, we always aim at maximum backward compatibility.

> Could you please describe your plans with iterable scanner?

To collect feedback on this issue and wait for somebody to be motivated to implement it. It's not on our short-term TODO list.

**Comment by pvojtechovsky** [comment]
CtIterableScanner might simplify algorithm of TemplateMatcher#helperMatch(...), which needs to synchronously iterate over both template and target.
Or different way around. We might extract algorithm of TemplateMatcher#helperMatch and use it to implement CtIterableScanner.

**Comment by AndreFCruz** [comment]
hey!
As a follow-up to #1980 we thought the CtElement interface could extend the Iterable<CtElement>, as expressed by @monperrus, but I don't see where that could be implemented.
The iterator() (and forEach()) methods should have the same implementation for all subclasses, but they don't have a common superclass to implement these (and not repeat code).

Another approach would be to have a wrapper class receive a CtElement and implement Iterable<CtElement> as seen [here](https://github.com/msramalho/spoon/blob/5fb5351831c73213517725f9f8fba16ad0c257f8/src/main/java/spoon/reflect/visitor/IterableCtElementWrapper.java).

Yet another approach would be having a Decorator
class IterableCtElement implements CtElement, Iterable<CtElement>
and just call every element of CtElement on its implementation, for ease of use.

Any thoughts?

**Comment by monperrus** [comment]
One lightweight solution is to simply add an "iterator" method in CtElement which would return a
CtIterator.

This is not exactly this feature but it's a good step.

**Comment by AndreFCruz** [comment]
But where would such a method be implemented?
As far as I can tell the code would be the same for all subclasses
```java
return new CtIterator(this);
```
This would have to be repeated on all subclasses, no?

**Comment by monperrus** [comment]
This can be only put once in CtElementImpl

**Comment by AndreFCruz** [comment]
You're right, my bad. The javadocs only listed the direct subclasses so I missed that all Impl subclasses eventually trace back to CtElementImpl.

**Comment by AndreFCruz** [comment]
I don't think this feature can be implemented without changing other classes (which means it probably isn't worth implementing).
The ```CtStatementList``` interface already extends ```Iterable<CtStatement>```, which creates conflicts if CtElement were to implement the Iterable<CtElement> interface, or even Iterable<E> (E extends CtElement).
Even with only the ```iterator()``` method the CtStatementList would have conflicting return types for the same method.

We get conflicts with ```CtStatementList``` with either one of these methods in CtElement:
```java
<E extends CtElement> Iterator<E> iterator();
```
or
```java
Iterator<CtElement> iterator();
```

**Comment by monperrus** [comment]
Thanks a lot for the analysis.

I still that it's a good idea to have a method in CtElement to help users find CtIterator.

What do you think of `Iterator CtElement#descendantIterator()`?

**Comment by AndreFCruz** [comment]
```Iterator<CtElement> descendantIterator()``` ?
Sure thing, I'll post a PR.

**Comment by monperrus** [comment]
closed per #1986
